Pos domain testing

jtec opened the pull request:
Partly fixes https://github.com/jtec/prx/issues/42

Adds a test that computes the least-squares PVT solution using pseudoranges and broadcast corrections and checks that the position solution is reasonably close to the approximate receiver position in the PRX file (which is based on the RINEX header).

The test is run using all observations as well as only observations of individual constellations to reveal possible constellatio-specific issues.

Some unrelated changes are caused by running `black .` .

jtec commented on src/prx/test/test_main.py:
```diff
@@ -68,7 +70,61 @@ def test_prx_function_call_with_csv_output(input_for_test):
+        solution_increment_sos = np.linalg.norm(x_lsq)
+        x_linearization += x_lsq
+        n_iterations += 1
+        assert n_iterations < 10
```
@plutonheaven Didn't get this to converge so far - do you see anything wrong with the way the LSQ solution is set up?

plutonheaven commented:
I corrected a bug.
The LSQ should be applied on $\mathbf{y - h(x0)}$, and you forgot to subtract $\mathbf{h(x0)}$.
Also, I changed some signs on the correction applications, which I think are the correct ones.
Finally, I added a simpler test, with just GPS C1C observations.

jtec commented:
> I corrected a bug. The LSQ should be applied on y−h(x0), and you forgot to subtract h(x0). Also, I changed some signs on the correction applications, which I think are the correct ones. Finally, I added a simpler test, with just GPS C1C observations.

Ahhhh of course - thanks!! I'll do some cleaning up and mark this ready for review.

plutonheaven commented on src/prx/test/test_main.py:
```diff
@@ -68,7 +71,98 @@ def test_prx_function_call_with_csv_output(input_for_test):
+    return records, metadata
+
+
+def spp_lsq(df):
```
Those functions (csv parser and spp_lsq) are quite useful, it kind of pains me to see them as additional functions in a `test_*.py` file.

Although they are not at the heart of **prx**, we may still put them forward in a separate `prx_use_illustration.py`, or something like that?

jtec replied:
Good call - I moved them into a separate `user.py` module :)

plutonheaven commented on src/prx/test/test_main.py:
```diff
@@ -68,7 +71,98 @@ def test_prx_function_call_with_csv_output(input_for_test):
+        df.time_of_reception_in_receiver_time
+        == df.time_of_reception_in_receiver_time.min()
+    ]
+    for constellations_to_use in [("G", "E", "C"), ("G",), ("E",), ("C",)]:
```
Great test execution!

plutonheaven commented on src/prx/main.py:
```diff
@@ -447,9 +446,17 @@ def process(observation_file_path: Path, output_format="jsonseq"):
+    records = build_records(
+        rinex_3_obs_file,
+        aux_files["broadcast_ephemerides"],
+        metadata["approximate_receiver_ecef_position_m"],
```
I see that this `approximate_receiver_ecef_position_m` is useful in our case (processing observations from fixed station), but I still think that a more detailed receiver position is needed, per epoch.

This is a separate issue, that I detail in #60.

So let us keep this as it is for this PR, but begin to think about more complex cases;

plutonheaven replied:
I just checked, the RNX header field `APPROX POSITION XYZ` is an optional field...

jtec replied:
Yes, thanks for opening https://github.com/jtec/prx/issues/60, it will be great to make that more robust.
